### PR TITLE
Several ASN decoder fixes

### PR DESCRIPTION
**Pull request by guidovranken**
See ZD 12349

### CODE REVIEW

**Comment by haydenroche5** [comment]
Hi @guidovranken thanks for the report and PR. The length checks look good to me. As for the CertStatus fix, I think we want to do something like this instead:

```
@@ -17470,6 +17474,9 @@ static int DecodeResponseData(byte* source,
             }
             single = single->next;
             XMEMSET(single, 0, sizeof(OcspEntry));
+            single->status = (CertStatus*)XMALLOC(sizeof(CertStatus),
+                resp->heap, DYNAMIC_TYPE_OCSP_STATUS);
+            XMEMSET(single->status, 0, sizeof(CertStatus));
             single->isDynamic = 1;
         }
     }
```

Each single response should have its own `CertStatus`. Adding @ejohnstown for further review.

**Review by haydenroche5 — APPROVED** [review]

**Review by SparkiDev — APPROVED** [review]

**Review by ejohnstown — APPROVED** [review]
A/B comparison through the fuzzer shows it is fixed. Good to go!

File: wolfcrypt/src/asn.c [hunk 1]
```diff
@@ -17470,6 +17474,9 @@ static int DecodeResponseData(byte* source,
             }
             single = single->next;
             XMEMSET(single, 0, sizeof(OcspEntry));
+            single->status = (CertStatus*)XMALLOC(sizeof(CertStatus),
```
Comment (dgarske) [inline wolfcrypt/src/asn.c:17477]: Thank you @bernd-edlinger for pointing out the malloc is not checked for NULL. 
This should have been caught in the multiple peer reviews....

I do not see this fixed in master or a pending PR. @ejohnstown or @guidovranken please put up a PR to fix.

Thanks, David Garske, wolfSSL

Comment (guidovranken) [reply wolfcrypt/src/asn.c:17477]: Fix: https://github.com/wolfSSL/wolfssl/pull/4093

**Pull request merged**